### PR TITLE
fix: error serialize circular references

### PR DESCRIPTION
**Pull request by 11joselu**
When there are a circular references in a object (like added test), serializeError enters in a infinite loop

image

image

This fix prevents to hit the maximum call stack

Relates to https://github.com/vitest-dev/vitest/issues/284

### CODE REVIEW

File: packages/vitest/src/runtime/error.ts [hunk 1]
```diff
@@ -2,14 +2,22 @@ import { format } from 'util'
...
+export function serializeError(val: any, seen = new WeakSet()): any {
   if (!val || typeof val === 'string')
     return val
 
   if (typeof val === 'function')
     return `Function<${val.name}>`
   if (typeof val !== 'object')
     return val
+
+  if (val !== null) {
```
Comment (userquin) [inline packages/vitest/src/runtime/error.ts:14]: the if statement is not necessary

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts]: Yes!. Removed :)

File: packages/vitest/src/runtime/error.ts [hunk 2]
```diff
@@ -2,14 +2,20 @@ import { format } from 'util'
...
+export function serializeError(val: any, seen = new WeakSet()): any {
   if (!val || typeof val === 'string')
     return val
 
   if (typeof val === 'function')
     return `Function<${val.name}>`
   if (typeof val !== 'object')
     return val
+
+  if (seen.has(val))
```
Comment (userquin) [inline packages/vitest/src/runtime/error.ts:14]: the check should be before the iteration, I'm testing on local and it seems with this approach the memory error still fails

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts:14]: With the test or another tkind of error?

I'm still fixing a Maximum Call error for another errors.

Comment (userquin) [reply packages/vitest/src/runtime/error.ts:14]: with `test:all`, testing mocks it works, I'll include your test to check it: 
imagen

Comment (userquin) [reply packages/vitest/src/runtime/error.ts:14]: your test also working, I see some invalid WeakSet usage error on some test on `test:ci`:

imagen

imagen

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts:14]: Sorry but I do not understand you. 

Is throwing an error runing the new test on your machine or we are talking about the pipeline checks?

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts:14]: Runing scripts on Linux Mint (ubuntu base)

-  pnpm test:ci -> Success 
- pnpm test:all -> Success

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts:14]: And re-executing throws this:

```
test/mocks test: Failed
undefined
/home/jcabrera/Desktop/vitest/test/mocks:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @vitest/test-lit@ test: `vitest`
Exit status 134
test/react-mui test$ vitest
```

Is that  error? Maybe is the same already commented on discord https://discord.com/channels/917386801235247114/917386801235247117/924403266001993788

Comment (userquin) [reply packages/vitest/src/runtime/error.ts:14]: Yes, it is the same. The error about invalid `WeakSet` usage error was when running `pnpm run test:ci`, your test was not included.

I run all tests using `nr ci` or `pnpm run test:ci` or `pnpm run test:all`: fail 5/6 times with the memory error on discord. In some execution, with your changes on `error.ts` I get invalid `WeakSet` usage error; then I move the check before the `forEach` and the invalid usage error disappear.

I also test each project running `pnpm run test` inside each directory: previous screenshots are the result moving the check before the `forEach` on `mocks` and on `core` with your test also included.

Comment (11joselu) [reply packages/vitest/src/runtime/error.ts]: Can you please tell me your OS and node + pnpm version? Maybe I can reproduce it

Comment (userquin) [reply packages/vitest/src/runtime/error.ts]: I have nvm with node 16.13.0 lts on Windows 10 pro with i9900k proccesor, pnpm latest version

**Pull request merged**